### PR TITLE
[resgen] Implement conditional resources (#if/#ifdef)

### PR DESCRIPTION
**Pull request by saper**
- Add support for #if, #ifdef, #if !SYMBOL, #else
and #endif directives

- Add /DEFINE: (/DEFINE:) option:

  /D:SYMBOL1 /D:SYMBOL2 or /DEFINE:SYMBOL1,SYMBOL2

This is needed to compile mscorlib resources
from CoreCLR

### CODE REVIEW

**Comment by alexrp** [comment]
The patch doesn't follow the Mono coding style.

File: mcs/tools/resgen/monoresgen.cs [hunk 1]
```diff
@@ -22,6 +23,7 @@ class ResGen {
 	static Assembly swf;
 	static Type resxr;
 	static Type resxw;
+	static List<string> symbols = new List<string> (20);
```
Comment (marek-safar) [inline mcs/tools/resgen/monoresgen.cs]: No need for explicit size and the allocation should be deferred

File: mcs/tools/resgen/monoresgen.cs [hunk 2]
```diff
@@ -74,7 +87,7 @@ class ResGen {
 			return new PoResourceReader (stream);
 		case ".txt":
 		case ".text":
-			return new TxtResourceReader (stream);
+			return new TxtResourceReader (stream, symbols.ToArray());
```
Comment (marek-safar) [inline mcs/tools/resgen/monoresgen.cs]: Why do you need an array

File: mcs/tools/resgen/monoresgen.cs [hunk 3]
```diff
@@ -215,6 +228,22 @@ class ResGen {
...
+				    args [i].StartsWith ("-d:")       ||
+				    args [i].StartsWith ("/define:")  ||
+				    args [i].StartsWith ("-define:")  ||
+				    args [i].StartsWith ("/D:")       ||
+				    args [i].StartsWith ("-D:")       ||
+				    args [i].StartsWith ("/DEFINE:")  ||
+				    args [i].StartsWith ("-DEFINE:")) {
+
+					foreach (string s in
+					         args [i].Substring (args [i].IndexOf (':') + 1).Split (',') ) {
```
Comment (marek-safar) [inline mcs/tools/resgen/monoresgen.cs]: use local variable for better readability

File: mcs/tools/resgen/monoresgen.cs [hunk 4]
```diff
@@ -358,10 +387,16 @@ static string Escape (string value)
...
 	Hashtable data;
 	Stream s;
+	Dictionary <String,int> defines;
 	
-	public TxtResourceReader (Stream stream) {
+	public TxtResourceReader (Stream stream, String[] symbols) {
 		data = new Hashtable ();
 		s = stream;
+
+		defines = new Dictionary<String, int> (symbols.Length);
```
Comment (marek-safar) [inline mcs/tools/resgen/monoresgen.cs]: Use HashSet instead

File: mcs/tools/resgen/monoresgen.cs [hunk 5]
```diff
@@ -371,17 +406,86 @@ class TxtResourceReader : IResourceReader {
 	public IDictionaryEnumerator GetEnumerator() {
 		return data.GetEnumerator ();
 	}
-	
+
+	public static string NextWord(ref string line) {
```
Comment (marek-safar) [inline mcs/tools/resgen/monoresgen.cs]: It does not need to be public

File: mcs/tools/resgen/monoresgen.cs [hunk 6]
```diff
@@ -371,17 +406,86 @@ class TxtResourceReader : IResourceReader {
...
+		int i, j;
+		string keywd;
+		line = line.TrimStart ();
+		for (i = 0; i < line.Length && !Char.IsWhiteSpace (line [i]) && line [i] != ';'; )
+			i++;
```
Comment (marek-safar) [inline mcs/tools/resgen/monoresgen.cs]: Move i++ after last ;

File: mcs/tools/resgen/monoresgen.cs [hunk 7]
```diff
@@ -371,17 +406,86 @@ class TxtResourceReader : IResourceReader {
...
+		return keywd;
+	}
+
 	void Load () {
 		StreamReader reader = new StreamReader (s);
 		string line, key, val;
+		Stack<bool> conditional = new Stack<bool> (5);
 		int epos, line_num = 0;
+
+		conditional.Push(true);
```
Comment (marek-safar) [inline mcs/tools/resgen/monoresgen.cs]: I'd prefer not to always push but do the check in endif

Comment (saper) [inline mcs/tools/resgen/monoresgen.cs]: But in that case checking for `conditional.Peek () && stat` gets unnecessarily complicated.. my first implementation just pushed true or false and used `conditional.contains(false)`

**Comment by saper** [comment]
Marek, thanks for your review, this is my first C# code :) I think I have addressed everything except for  the Push thing. We can have multiple `/DEFINE:` options, so I think `ResGen.symbols` initialization at the top will be the best...

File: mcs/tools/resgen/monoresgen.cs (hunk 6 above)
Comment (saper) [inline mcs/tools/resgen/monoresgen.cs]: Like this?

     for(i = 0; ....; i++);

Do you have any standard for empty `for` loops? Easy to miss semicolon at the end...

**Comment by atsushieno** [comment]
Unless there is any reason to not to, I will merge this. I also need this feature.

**Comment by jasonwilliams200OK** [comment]
@atsushieno, can this land in next release of mono?

**Comment by atsushieno** [comment]
Oh yes! Let's go ahead and do so.

**Pull request merged**